### PR TITLE
expose the rust cargo.toml version string

### PR DESCRIPTION
**Pull request by fundthmcalculus**

### CODE REVIEW

File: native/src/ffi/utils.rs [hunk 1]
```diff
@@ -45,3 +47,13 @@ pub unsafe extern "C" fn okapi_string_free(s: *mut std::os::raw::c_char) {
...
     });
 }
+
+#[no_mangle]
+pub extern "C" fn okapi_version() -> *const c_char {
+    // TODO - Insert git commit ID
+    const PKG_VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");
+    let rust_str = CString::new(PKG_VERSION.unwrap()).unwrap();
+    let c_ver = rust_str.as_ptr();
+    std::mem::forget(rust_str);
```
Comment (tmarkovski) [inline native/src/ffi/utils.rs:57]: This part leaks memory, as there's no way to clear it

Comment (fundthmcalculus) [reply native/src/ffi/utils.rs:57]: We call back into `okapi_string_free` to free it client side. Hmmm.

**Comment by tmarkovski** [comment]
The method definition can be the same as other FFI methods

```
int32_t okapi_metadata(struct ByteBuffer request,
                       struct ByteBuffer *response,
                       struct ExternError *err);
```

```
message MetadataRequest {
   repeated variables = 1; // optional field, can contain any of the cargo env vars
}
```

```
message MetadataResponse {
  string version = 1;
  int version_major = 2;
  int version_minor = 3;
  // etc
  map<string, string> variables = 10; // will include any non default requested variables
}
```

**Comment by fundthmcalculus** [comment]
That makes a lot of sense. Let's go with what you described. Is there a reason to return the full version in addition to the major and minor numbers? I think we should just return the version string and leave it at that. Definitely have the metadata map though.

**Comment by fundthmcalculus** [comment]
> Because it's simple to do so, and wrapper may make decisions based on different major versions, but not care about minor or patch changes for example.

That does make a good deal of sense. Since we aren't really releasing patch versions of the okapi binaries much, that should be enough.

**Comment by fundthmcalculus** [comment]
> * Getting version should be expanded to get [additional metadata](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) from native, not just version

These are only available at compile-time, so custom requesting of the information is not possible. `env!()` and `option_env!()` require a compile-time string constant, not a variable.

File: native/Cargo.toml [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+authors = ["Tomislav Markovski <tomislav@trinsic.id>", "Scott Phillips <scott.phillips@trinsic.id>"]
 edition = "2018"
 # ensure this is not checked in as "true" as it will
 # cause vendored libraries to fail compilation due
 # to dependency on protoc compiler
 # if you need to compile .proto files, use this line:
-#build = "build.rs"
-build = false
+build = "build.rs"
+#build = false
```
Comment (tmarkovski) [inline native/Cargo.toml:11]: This needs be false and not commented out. The reason for that is that it makes NixOS builds fail.
A workaround would be to remove this file before packaging for nixos, but for now, let's just keep it to false when checking in.

**Review by tmarkovski — APPROVED** [review]
Can you do `cargo fmt`? It seems there's a file that needs formatting.

File: native/src/metadata/mod.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        let env_vars = request.variables.clone();
+
+        // Get the version for sure, and parse it
+        let mut major_ver = 0;
+        let mut minor_ver = 0;
+
+        const PKG_VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");
+        let pkg_pieces = PKG_VERSION.unwrap().split(".").collect::<Vec<&str>>();
+        if pkg_pieces.len() >= 2 {
+            major_ver = pkg_pieces[0].parse::<i32>().unwrap();
```
Comment (tmarkovski) [inline native/src/metadata/mod.rs:16]: There's a specific cargo variable to return this, don't parse it from the full version.

File: native/src/metadata/mod.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+use crate::didcomm::Error;
+use crate::proto::okapi::okapi_metadata::{MetadataRequest, MetadataResponse};
+
+impl crate::Metadata {
+    pub fn get_metadata<'a>(request: &MetadataRequest) -> Result<MetadataResponse, Error<'a>> {
+        let env_vars = request.variables.clone();
+
+        // Get the version for sure, and parse it
+        let mut major_ver = 0;
+        let mut minor_ver = 0;
```
Comment (tmarkovski) [inline native/src/metadata/mod.rs:11]: Also include patch please. All are available as cargo vars. See link in my comment.

File: native/src/metadata/mod.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+        const PKG_VERSION: Option<&str> = option_env!("CARGO_PKG_VERSION");
+        let pkg_pieces = PKG_VERSION.unwrap().split(".").collect::<Vec<&str>>();
```
Comment (tmarkovski) [inline native/src/metadata/mod.rs:14]: `unwrap` can panic here, use ? to propagate error or assign default value to it using `unwrap_or`

File: native/Cargo.toml (hunk 2 above)
Comment (fundthmcalculus) [reply native/Cargo.toml:11]: Yup, this was an oversight on my part. Thanks for catching this!

File: native/src/metadata/mod.rs (hunk 5 above)
Comment (fundthmcalculus) [reply native/src/metadata/mod.rs]: Decided on `unwrap_or_default()` since a blank is the default for not found per their docs.

**Review by tmarkovski — APPROVED** [review]

**Comment by fundthmcalculus** [comment]
I will handle rubocop and flake8 code style issues in a separate PR.

**Pull request merged**